django_keycloak/keycloak.py

```diff
@@ -338,17 +320,102 @@ def update_user(self, user_id, **values):
+        server_url, headers = self._make_secure_json_request_config()
+
+        url = KEYCLOAK_CREATE_USER.format(server_url, self.realm)
+        res = requests.post(url, headers=headers, json=values)
+        res.raise_for_status()
+        return res.headers["Location"].split("/")[-1]
+
+    @keycloak_api_error_handler
+    def delete_user(self, user_id):
+        server_url, headers = self._make_secure_json_request_config()
+
+        url = KEYCLOAK_DELETE_USER.format(server_url, self.realm, user_id)
+        res = requests.delete(url, headers=headers)
+        res.raise_for_status()
+
+    def _is_client_token_timed_out(self):
+        """
+        Checks if the cached token is timed out as given by the timeout factor
+
+        If there is no cached token, returns true. Timeout is true if the current time is
+        later than (expiration time - issue time) * timeout factor + issue time.
+        """
+        if not self._client_token:
+            return True
+
+        # Get the public key from the identity provider, i.e., the keycloak server
+        decoded = self._decode_token(self._client_token, self.client_jwt_audience)
+        exp = datetime.fromtimestamp(decoded["exp"], tz=timezone.utc)
+        iat = datetime.fromtimestamp(decoded["iat"], tz=timezone.utc)
+        now = datetime.now(tz=timezone.utc)
+
+        is_timed_out = (exp - iat) * self.token_timeout_factor + iat < now
+        return is_timed_out
+
+    def _get_openid_config(self):
+        server_url, headers = self._make_json_request_config()
+
+        response = requests.request(
+            "GET",
+            KEYCLOAK_OPENID_CONFIG.format(server_url, self.realm),
+            headers=headers,
+        )
+        return response.json()
 
+    def _make_secure_request_config(self, token):
+        server_url, headers = self._make_request_config()
+        if not token:
+            token = self.get_token()
+        headers["Authorization"] = f"Bearer {token}"
+        return [server_url, headers]
+
+    def _make_secure_json_request_config(self, token=None):
+        server_url, headers = self._make_request_config()
+        if not token:
+            token = self.get_token()
+        headers["Content-Type"] = "application/json"
+        headers["Authorization"] = f"Bearer {token}"
+        return [server_url, headers]
+
+    def _make_form_request_config(self):
+        server_url, headers = self._make_request_config()
+        headers["Content-Type"] = "application/x-www-form-urlencoded"
+        return [server_url, headers]
+
+    def _make_json_request_config(self):
+        server_url, headers = self._make_request_config()
+        headers["Content-Type"] = "application/json"
+        return [server_url, headers]
+    
+    def _make_request_config(self):
         server_url = self.server_url
+        headers = {}
         if self.internal_url:
             server_url = self.internal_url
             headers["HOST"] = urlparse(self.server_url).netloc
```

Comment:
When using `self.internal_url` the `HOST` header must be `urlparse(self.internal_url).netloc` and not `urlparse(self.server_url).netloc`. This is a bug that was already here, but noticed it when testing with the `DECODE_TOKEN=True`, `SERVER_URL="http://localhost:8080"` and `INTERNAL_URL=http://keycloak:8080`. The subsequent requests done by `public_key = self.jwks_client.get_signing_key_from_jwt(token).key ` used `http://localhost:8080` instead of http://keycloak:8080 and since everything was inside docker the request could not found `http://localhost:8080`. 

Maybe we can simplify this and have only one URL that can be publicly accessible or an internal docker network reference (future improvement, I think).

Reply:
Makes sense. I've also just looked through the code and when internal URL is used, it overrides all uses of server_url, thereby supporting your idea to simplify the configuration to only use one url setting